zk-token-sdk/src/curve25519/curve_syscall_traits.rs

```diff
@@ -0,0 +1,98 @@
+//! zk-token-sdk or curve25519. It should be moved to a more general location in the future.
+//!
+
+pub trait PointValidation {
+    type Point;
+
+    /// Verifies if a byte representation of a curve point lies in the curve.
+    fn validate_point(&self) -> bool;
+}
+
+pub trait GroupOperations {
+    type Point;
+    type Scalar;
+
+    /// Adds two curve points: P_0 + P_1.
+    fn add(left_point: &Self::Point, right_point: &Self::Point) -> Option<Self::Point>;
+
+    /// Subtracts two curve points: P_0 - P_1.
+    ///
+    /// NOTE: Altneratively, one can consider replacing this with a `negate` function that maps a
+    /// curve point P -> -P. Then subtraction can be computed by combining `negate` and `add`
+    /// syscalls. However, `subtract` is a much more widely used function than `negate`.
+    fn subtract(left_point: &Self::Point, right_point: &Self::Point) -> Option<Self::Point>;
+
+    /// Multiplies a scalar S with a curve point P: S*P
+    fn multiply(scalar: &Self::Scalar, point: &Self::Point) -> Option<Self::Point>;
+}
+
+pub trait MultiScalarMultiplication {
+    type Scalar;
+    type Point;
+
+    /// Given a vector of scalsrs S_1, ..., S_N, and curve points P_1, ..., P_N, computes the
+    /// "inner product": S_1*P_1 + ... + S_N*P_N.
+    ///
+    /// NOTE: This operation can be represented by combining `add` and `multiply` functions in
+    /// `GroupOperations`, but computing it in a single batch is significantly cheaper. Given how
+    /// commonly used the multiscalar multiplication (MSM) is, it seems to make sense to have a
+    /// designated trait for MSM support.
+    ///
+    /// NOTE: The inputs to the function is a non-fixed size vector and hence, there are some
+    /// complications in computing the cost for the syscall. The computational costs should only
+    /// depend on the length of the vectors (and the curve), so it would be ideal to support
+    /// variable length inputs and compute the syscall cost as is done in eip-197:
+    /// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-197.md#gas-costs. If not, then we can
+    /// consider bounding the length of the input and assigning worst-case cost.
+    fn multiscalar_multiply(
+        scalars: Vec<&Self::Scalar>,
+        points: Vec<&Self::Point>,
+    ) -> Option<Self::Point>;
+}
+
+pub trait Pairing {
+    type G1Point;
+    type G2Point;
+    type GTPoint;
+
+    /// Applies the bilinear pairing operation to two curve points P1, P2 -> e(P1, P2). This trait
+    /// is only relevant for "pairing-friendly" curves such as BN254 and BLS12-381.
+    fn pairing_map(
+        left_point: &Self::G1Point,
+        right_point: &Self::G2Point,
+    ) -> Option<Self::GTPoint>;
+}
+
+pub const CURVE25519_EDWARDS: u64 = 0;
+pub const CURVE25519_RISTRETTO: u64 = 1;
+// BN254 curve to be instantiated in the future
+/* pub const BN254_G1: u64 = 2;
+pub const BN254_G2: u64 = 3;
+pub const BN254_GT: u64 = 4; */
```

Comment:
should we avoid explicit curve25519 naming throughout if we plan to add other curves in the future?

Reply:
Yeah, so the traits are named independent of curve25519. The sycalls are also currently named independent of curve25519: `SyscallValidatePoint::call` should take in a `curve_id` as the first parameter that should identify the curve function to run.

The feature set is named specifically for curve25519 since the plan is to support curve25519 first. Then I was hoping we can have another feature set that supports another curve afterwards that extends the functoin call again. I don't know if this is possible/correct thing to do though...